pkgs/applications/terminal-emulators/mlterm/default.nix

```diff
@@ -1,14 +1,82 @@
-{ stdenv, lib, fetchFromGitHub, pkg-config, autoconf, makeDesktopItem, nixosTests
-, libX11, gdk-pixbuf, cairo, libXft, gtk3, vte
-, harfbuzz #substituting glyphs with opentype fonts
-, fribidi, m17n_lib #bidi and encoding
+{ stdenv
+, lib
+, fetchFromGitHub
+, pkg-config
+, autoconf
+, makeDesktopItem
+, nixosTests
+, vte
+# Open Type layout support, (substituting glyphs with opentype fonts)
+, enableOtl ? true
+, harfbuzz # can be replaced with libotf
+, enableBidi ? true
+, fribidi
+, enableM17n ? !stdenv.isDarwin
+, m17n_lib
+, enableSsh2 ? true
 , libssh2 #build-in ssh
-, fcitx5, fcitx5-gtk, ibus, uim #IME
+, enableFcitx ? !stdenv.isDarwin
+, fcitx5, fcitx5-gtk
+, enableIbus ? !stdenv.isDarwin
+, ibus
+, enableUim ? !stdenv.isDarwin
+, uim #IME
 , wrapGAppsHook #color picker in mlconfig
-, Cocoa #Darwin
+, imagelibChoice ? "gdk-pixbuf" # or mlimgloader depending on your bugs of choice
+, gdk-pixbuf
+, gtk3
+, gtk ? gtk3
+# List of gui libraries to use. According to `./configure --help` ran on
+# release 3.9.3, options are: (xlib|win32|fb|quartz|console|wayland|sdl2|beos)
+, guiChoices ? [
+  "xlib"
+  "fb"
+]
+, xquartz # for the "quartz" --with-gui option
+, libxkbcommon , wayland # for the "wayland" --with-gui option
+, SDL2 # for the "sdl" --with-gui option
+# List of typing engines, the default list enables compiling all of the
+# available ones, as recorded on release 3.9.3
+, typeEnginesChoices ? [
+  "xcore"
+  "xft"
+  "cairo"
+]
+, libX11
+, libXft
+, cairo
+# List of external tools to create, this default list includes all default
+# tools, as recorded on release 3.9.3. mlterm-menu and mlconfig depend on
+# enabling gnome.at-spi2-core and configuring ~/.mlterm/key correctly.
+, toolsChoices ? [
+  "mlclient"
+  "mlconfig"
+  "mlcc"
+  "mlterm-menu"
+  "mlimgloader"
+  "registobmp"
+  "mlfc"
+]
+# Whether to enable the X window system
+, enableX ? true
+# Most of the input methods and other build features are enabled by default,
+# the following attribute set can be used to disable some of them. It's parsed
+# when we set `configureFlags`.
+, extraEnableFlags ? {
+  /* The format is as follows:
+  dnd = true;
+  iiimf = false;
+  scim = false;
+  */
+}
```

Comment:
Neat feature but many flags probably need extra deps aswell and at that point you're in overrideAttrs territory anyways.

I like this pattern though as, contrary to list args, set args can work well with overrides because sets can be updated and attrs thereby overridden. I.e. `mlterm.override (old: { extraEnableFlags = old.extraEnableFlags // { foo = false, bar = true; }; })`

Could also be a good alternative to your feature lists.

Reply:
> Neat feature but many flags probably need extra deps aswell

I haven't thoroughly checked, but I kind of got the impression that most if not all of them don't.

> I like this pattern though as, contrary to list args

Me too, and one of the best implementations I had for this kind of Nix build configuration is in Gnuradio:

https://github.com/NixOS/nixpkgs/blob/2ba48cc8dee2249526efeda1d2c629a2ec625229/pkgs/applications/radio/gnuradio/default.nix#L42-L85

However, here it'll become a little more complicated, as the naming of the options in `./configure` are not consistent as in Gnuradio... If you are up to these idea, I'll find time to go ahead and implement it.

Reply:
Friendly ping for @Atemu regarding the suggestion for a GNURadio like function arguments API.

Reply:
BTW, regarding the `.override` API: Even GNUradio's `features ? {}` argument doesn't allow to add / delete buildInputs directly via `.override`. I think it's too much to ask from such an interface. Also, if a feature requires a certain dependency not written in Nixpkgs, the minimum I'd expect from such a Nix user is to contribute this change. If they don't want to, they can always return to `.overrideAttrs`.

I do agree regarding the list vs attribute set though. I'll change this today, and perhaps put all current `enableThisFeature` flags there as a default value.

Reply:
Done.

Reply:
> Even GNUradio's `features ? {}` argument doesn't allow to add / delete buildInputs directly via `.override`. I think it's too much to ask from such an interface.

Sure. My point is that because free-form GNUradio-style feature enablement doesn't give you as much freedom as you'd think, having it free-form over fixed `enableFooFeature` arguments is less of a benefit. I think it might still win out though. Perhaps I'll refactor ffmpeg's fixed feautre flag flood to this pattern too.